### DISCUSSION TITLE
VideoWritter(ffmpeg) makes strange artefact with mono 12 bits video input

### DISCUSSION THREAD

**Discussion by frr665** [bonsai-rx/bonsai#1681] [Q&A]
Hello Bonsai community,

I'm encountering an issue with a simple script that I'm using to process video input from Pylon using the `PylonCapture` node. The script involves cropping the video into quarters and saving the cropped video as an .avi file using `VideoWriter (ffmpeg)`. 
![bonsai_12bit_snapshot](https://github.com/bonsai-rx/bonsai/assets/117995542/96a0b592-5cec-4978-9b4e-dd6562e4caeb)

Everything works smoothly when the input video is 8-bit mono, however, I'm facing strange visual artifacts when the input video is mono 12-bit. 
Here are the ffmpeg arguments I'm currently using: 
`-pix_fmt gray12le -c:v ffv1 -level 3 -coder 1 -context 1 -g 1 -slices 24 -slicecrc 1 -top 0 -an`.

PixelFormat	Mono8 : everything is fine
<img src="https://github.com/bonsai-rx/bonsai/assets/117995542/d1201ec8-921e-4400-9a22-e2c664994844" alt="image" width="50%" height="auto">

PixelFormat	Mono12 : Here, the frame seems to be zoomed in and the video is roughly 3 time longer than the real recording time. 
<img src="https://github.com/bonsai-rx/bonsai/assets/117995542/1cec0d30-19b5-42ef-820a-c2667ef025eb" alt="image" width="50%" height="auto">

Interestingly, when I save the video input as TIFF images using `VideoWriter (Vision)` , the images appear normal. However, I prefer to have a compressed video file as the output instead of a sequence of images because I need to record really long video for this experiment. 

I would greatly appreciate any assistance or insights into resolving this issue. Thank you!

Hugo

Edit : 
here is my workflow and the pylon parameters file 
[test_12_bits.zip](https://github.com/bonsai-rx/bonsai/files/14362321/test_12_bits.zip)

**Comment by bruno-f-cruz** [comment]
Please add your workflow, thanks

**Reply by frr665** [reply]
Hi Bruno and thanks for answering, I edited my initial post to add my workflow.

**Comment by bruno-f-cruz** [comment]
I suspect it may be because of this line:

https://github.com/bonsai-rx/ffmpeg/blob/14757e1252ad9499e04c4f5399b5650c2c05fdf5/Bonsai.FFmpeg/VideoWriter.cs#L53C1-L54C1

It seems to force the format to be 8bit (grayscale or rgb). You can modify the script and use it in your Extensions folder or you can create the pipe and the ffmpeg process manually ([see this post](https://github.com/orgs/bonsai-rx/discussions/919#discussioncomment-2895689))

**Reply by frr665** [reply]
Ok thanks for your insights. I think that modifying this line of code for `image.Channels == 1 ? "gray12le" : "bgr24"` should do the trick.  
But how to enforce this modification ? Do I need to create a new package like described here https://bonsai-rx.org/docs/articles/create-package.html ? 
Or can I just clone the the source code of `VideoWriter (ffmpeg)` in extension and modify it directly ?

**Reply by bruno-f-cruz** [reply]
https://bonsai-rx.org/docs/articles/scripting-extensions.html

You should be able to use it as an extension script
